lib/kdl/include/kdl/expected.h

```diff
@@ -0,0 +1,627 @@
+            return absl::visit(
+                std::forward<Visitor>(visitor),
+                std::move(expected.m_value));
+        }
+        
+        /**
+         * Applies the given function to the contained value and returns a new expectation with the result type of the
+         * given function as its value type.
+         *
+         * If this expectation contains a value, the function is applied to it and the result of applying the function
+         * is returned wrapped in an expectation.
+         * If this expectation contains an error, that error is returned as is.
+         *
+         * The value is passed to the mapping function by const lvalue reference.
+
+         * @tparam F the type of the function to apply
+         * @param f the function to apply
+         */
+        template <typename F>
+        auto map(F&& f) const & {
+            using R = std::invoke_result_t<F, Value>;
+            return visit(kdl::overload {
+                [&](const value_type& v) { return expected<R, Errors...>(f(v)); },
+                [] (const auto& e)       { return expected<R, Errors...>(e); }
+            }, *this);
+        }
+        
+        /**
+         * Applies the given function to the contained value and returns a new expectation with the result type of the
+         * given function as its value type.
+         *
+         * If this expectation contains a value, the function is applied to it and the result of applying the function
+         * is returned wrapped in an expectation.
+         * If this expectation contains an error, that error is returned as is.
+         *
+         * The value is passed to the mapping function by rvalue reference.
+         *
+         * The given function can move the value out of this expectation, so care must be taken not to
+         * use this expectation afterwards, as it will be in a moved-from state.
+
+         * @tparam F the type of the function to apply
+         * @param f the function to apply
+         */
+        template <typename F>
+        auto map(F&& f) && {
+            using R = std::invoke_result_t<F, Value>;
+            return visit(kdl::overload {
+                [&](value_type&& v) { return expected<R, Errors...>(f(std::move(v))); },
+                [] (const auto& e)  { return expected<R, Errors...>(e); }
+            }, *this);
+        }
+
+        /**
+         * Indicates whether this expectation contains a value.
+         */
+        bool success() const {
+            return absl::holds_alternative<value_type>(m_value);
+        }
+        
+        /**
+         * Indicates whether this expectation contains an error.
+         */
+        bool error() const {
+            return !success();
+        }
+        
+        /**
+         * Indicates whether this expectation contains a value.
+         */
+        operator bool() const {
+            return success();
+        }
+    };
+    
+    /**
```

Comment:
I followed along up to this point, but I don't really understand why we need any of the further types.
Also it means they have to reimplement all of kdl::expected, port bugfixes between all the copied code, etc.

If you want to return a `int&` why not just make a type `kdl::expected<std::reference_wrapper<int>, ...>` in the calling code?

Reply:
Because this is easier to read and understand (IMO) for users of this class. It's true that it adds a burden of maintenance to the library, but I think it makes using these types far easier, since `result<int&, This, That>` expresses intent much better than `result<std::reference_wrapper<int>, This, That>`.

Same goes for `result<void, This, That>` vs. `result<success_tag, This, That>`, but I am not sure about `result<opt<int>, This, That>` vs. `result<std::optional<int>, This, That>`. I guess in the latter case, using `std::optional` instead of `opt` is clearer, however I like that a visitor of an optional result value can take a `int` directly without having to care whether the value actually exists.

Reply:
Ah I see the goal of the `opt<>` one now. Although it feels like it could be implemented as a special `visit()` override on the basic `result<T>`, rather than a whole new class - it could even have a different name like `visit_and_unwrap_optional()`.

I'll need to actually use these a bit to comment more.. There are a lot of ways `visit()` could work. I suspect for `result<opt<int>` we'll want `visit()` to run even when the result contains an empty optional, and for `result<void` we'll want visit to run when it's in the "success" (empty) state.  e.g. `visit()` could take two lambdas instead of just one, a success and a failure lambda.

> since result<int&, This, That> expresses intent much better than result<std::reference_wrapper<int>, This, That>.

It's syntax sugar, not sure about expressing intent better. Hopefully we don't return references enough to need syntax sugar for this, since it's dangerous, but I guess it is a nicer syntax.